api/src/main/java/io/grpc/CallOptions.java

```diff
@@ -41,42 +41,68 @@ public final class CallOptions {
   /**
    * A blank {@code CallOptions} that all fields are not set.
    */
-  public static final CallOptions DEFAULT = new CallOptions();
+  public static final CallOptions DEFAULT = new Builder().build();
 
-  // Although {@code CallOptions} is immutable, its fields are not final, so that we can initialize
-  // them outside of constructor. Otherwise the constructor will have a potentially long list of
-  // unnamed arguments, which is undesirable.
   @Nullable
-  private Deadline deadline;
-  
+  private final Deadline deadline;
+
   @Nullable
-  private Executor executor;
+  private final Executor executor;
 
   @Nullable
-  private String authority;
+  private final String authority;
 
   @Nullable
-  private CallCredentials credentials;
+  private final CallCredentials credentials;
 
   @Nullable
-  private String compressorName;
+  private final String compressorName;
 
-  private Object[][] customOptions;
+  private final Object[][] customOptions;
 
-  // Unmodifiable list
-  private List<ClientStreamTracer.Factory> streamTracerFactories = Collections.emptyList();
+  private final List<ClientStreamTracer.Factory> streamTracerFactories;
 
   /**
    * Opposite to fail fast.
    */
   @Nullable
-  private Boolean waitForReady;
+  private final Boolean waitForReady;
 
   @Nullable
-  private Integer maxInboundMessageSize;
+  private final Integer maxInboundMessageSize;
   @Nullable
-  private Integer maxOutboundMessageSize;
+  private final Integer maxOutboundMessageSize;
+
+  private CallOptions(Builder builder) {
+    this.deadline = builder.deadline;
+    this.executor = builder.executor;
+    this.authority = builder.authority;
+    this.credentials = builder.credentials;
+    this.compressorName = builder.compressorName;
+    this.customOptions = builder.customOptions;
+    this.streamTracerFactories = builder.streamTracerFactories;
+    this.waitForReady = builder.waitForReady;
+    this.maxInboundMessageSize = builder.maxInboundMessageSize;
+    this.maxOutboundMessageSize = builder.maxOutboundMessageSize;
+  }
 
+  static class Builder {
+    Deadline deadline;
+    Executor executor;
+    String authority;
+    CallCredentials credentials;
+    String compressorName;
+    Object[][] customOptions = new Object[0][2];
```

Comment:
Although this object will only be used once, this will create it all the time. This and the emptyList() below are just for DEFAULT, so we should just make a static method to configure the default builder, or use a static block. I'll do that as a follow-up.

The optimizer should optimize these away, but it would still be there on startup and cause a non-trivial constructor. So it isn't a big deal, but still worth doing.

For reference, it'll look something like this:

```
static {
  Builder b = new Builder();
  b.customOptions = new Object[0][2];
  b.streamTracerFactories = Collections.emptyList();
  DEFAULT = b.build();
}

// or

public static final CallOptions DEFAULT = newDefault();

private static CallOptions newDefault() {
  Builder b = new Builder();
  b.customOptions = new Object[0][2];
  b.streamTracerFactories = Collections.emptyList();
  return b.build();
}
```